server/kv.go

```diff
@@ -0,0 +1,694 @@
+			fmt.Sprintf("failed to load confluence user for mattermostUserId:%s", mattermostUserID))
+	}
+	return user, nil
+}
+
+func (store store) CountUsers() (int, error) {
+	count := 0
+	for i := 0; ; i++ {
+		keys, appErr := store.plugin.API.KVList(i, listPerPage)
+		if appErr != nil {
+			return 0, appErr
+		}
+
+		for _, key := range keys {
+			if strings.HasPrefix(key, prefixUser) {
+				count++
+			}
+		}
+
+		if len(keys) < listPerPage {
+			break
+		}
+	}
+	return count, nil
+}
+
+func (store store) MapUsers(f func(user *User) error) error {
+	for i := 0; ; i++ {
+		keys, appErr := store.plugin.API.KVList(i, listPerPage)
+		if appErr != nil {
+			return appErr
+		}
+
+		for _, key := range keys {
+			if !strings.HasPrefix(key, prefixUser) {
+				continue
+			}
+
+			user := NewUser("")
+			err := store.get(key, user)
+			if err != nil {
+				return errors.WithMessage(err, fmt.Sprintf("failed to load confluence user for key:%s", key))
+			}
+
+			err = f(user)
+			if err != nil {
+				return err
+			}
+		}
+
+		if len(keys) < listPerPage {
+			break
+		}
+	}
+	return nil
+}
+
+// MigrateV2User migrates a user record from the V2 data model if needed. It
+// returns an up-to-date User object either way.
+func (p *Plugin) MigrateV2User(mattermostUserID types.ID) (*User, error) {
+	user, err := p.userStore.LoadUser(mattermostUserID)
+	if errors.Cause(err) != kvstore.ErrNotFound {
+		// return the existing key (or error)
+		return user, err
+	}
+
+	// V3 "user" key does not. Migrate.
+	instances, err := p.instanceStore.LoadInstances()
+	if err != nil {
+		return nil, err
+	}
+
+	user = NewUser(mattermostUserID)
+	for _, instanceID := range instances.IDs() {
+		_, err = p.userStore.LoadConnection(instanceID, mattermostUserID)
+		if errors.Cause(err) == kvstore.ErrNotFound {
+			continue
+		}
+		if err != nil {
+			return nil, err
+		}
+		user.ConnectedInstances.Set(instances.Get(instanceID))
+	}
+	err = p.userStore.StoreUser(user)
+	if err != nil {
+		return nil, err
+	}
+
+	return user, nil
+}
+
+func (store store) StoreOAuth2State(state string) error {
+	// Expire in 15 minutes
+	appErr := store.plugin.API.KVSetWithExpiry(
+		hashkey(prefixOneTimeSecret, state), []byte(state), 15*60)
+	if appErr != nil {
+		return errors.WithMessage(appErr, "failed to store state "+state)
+	}
+	return nil
+}
+
+func (store store) VerifyOAuth2State(state string) error {
+	data, appErr := store.plugin.API.KVGet(hashkey(prefixOneTimeSecret, state))
+	if appErr != nil {
+		return errors.WithMessage(appErr, "failed to load state "+state)
+	}
+
+	if string(data) != state {
+		return errors.New("invalid oauth state, please try again")
+	}
+
+	return nil
+}
+
+func (store store) EnsureAuthTokenEncryptionSecret() (secret []byte, returnErr error) {
+	defer func() {
+		if returnErr == nil {
+			return
+		}
+		returnErr = errors.WithMessage(returnErr, "failed to ensure auth token secret")
+	}()
+
+	// nil, nil == NOT_FOUND, if we don't already have a key, try to generate one.
```

Comment:
not sure about this, shouldn't we trigger a more detailed error in that special case where both are nil? If we are returning the error, but the error is nil, I think it loses the purpose of returning the error, right?

Reply:
we checking this below like the first condition is if the error is not nil then return the error then we are checking whether the len(secret) == 0 then we are generating a new secret for that so here we won't be returning the error is both are nil.